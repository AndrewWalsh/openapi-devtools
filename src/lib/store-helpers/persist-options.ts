import store2 from "store2";
import type { Options } from "../RequestStore";

const namespace = "openapi-devtools-options";
const key = "options";
const store = store2.namespace(namespace);

export const defaultOptions: Readonly<Options> = {
  enableMoreInfo: false,
};

const get = (): Options => {
  const options = store.get(key);
  if (options) return options;
  store.set(key, defaultOptions);
  return defaultOptions;
};
const set = (value: Partial<Options>) => {
  const options = get();
  const newOptions = { ...options, ...value };
  store.set(key, newOptions);
};
const clear = () => store.clear();

export default {
  get,
  set,
  clear,
};
