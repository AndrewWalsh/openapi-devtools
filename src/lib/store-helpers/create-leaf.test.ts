import { it, expect } from "vitest";
import { defaultOptions } from "./persist-options";
import postJson from "../__fixtures__/post-application-json";
import postXWWWFormUrlEncoded from "../__fixtures__/post-application-x-www-form-urlencoded";
import createLeaf, { Params } from "./create-leaf";
import { cloneDeep } from "lodash";

it("creates a leaf when request mime starts with application/x-www-form-urlencoded;charset=UTF-8", () => {
  const harRequest = cloneDeep(postXWWWFormUrlEncoded);
  harRequest.request.postData!.mimeType = "application/x-www-form-urlencoded;charset=UTF-8";
  const responseBody: Params["responseBody"] = { test: 1 };
  const options = defaultOptions;
  const leaf = createLeaf({ harRequest, responseBody, options });
  expect(leaf).toMatchSnapshot();
});

it("creates a leaf when request mime = application/x-www-form-urlencoded", () => {
  const harRequest = postXWWWFormUrlEncoded;
  const responseBody: Params["responseBody"] = { test: 1 };
  const options = defaultOptions;
  const leaf = createLeaf({ harRequest, responseBody, options });
  expect(leaf).toMatchSnapshot();
});

it("creates a leaf when mime = application/json;something for req and res", () => {
  const harRequest = cloneDeep(postJson);
  harRequest.request.postData!.mimeType = "application/json;something";
  const responseBody: Params["responseBody"] = { test: 1 };
  const options = defaultOptions;
  const leaf = createLeaf({ harRequest, responseBody, options });
  expect(leaf).toMatchSnapshot();
});

it("creates a leaf when mime = application/json for req and res", () => {
  const harRequest = postJson;
  const responseBody: Params["responseBody"] = { test: 1 };
  const options = defaultOptions;
  const leaf = createLeaf({ harRequest, responseBody, options });
  expect(leaf).toMatchSnapshot();
});
