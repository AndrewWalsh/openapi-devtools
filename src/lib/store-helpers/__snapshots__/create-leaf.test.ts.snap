// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`creates a leaf when mime = application/json for req and res 1`] = `
{
  "methods": {
    "POST": {
      "queryParameters": {
        "properties": {
          "alt": {
            "type": "string",
          },
          "key": {
            "type": "string",
          },
        },
        "required": [
          "alt",
          "key",
        ],
        "type": "object",
      },
      "request": {
        "application/json": {
          "body": {
            "properties": {
              "test": {
                "type": "string",
              },
            },
            "required": [
              "test",
            ],
            "type": "object",
          },
        },
      },
      "requestHeaders": {
        "properties": {
          "anonymousid": {
            "type": "string",
          },
        },
        "required": [
          "anonymousid",
        ],
        "type": "object",
      },
      "response": {
        "200": {
          "application/json": {
            "body": {
              "properties": {
                "test": {
                  "type": "integer",
                },
              },
              "required": [
                "test",
              ],
              "type": "object",
            },
          },
        },
      },
      "responseHeaders": undefined,
    },
  },
  "pathname": "/v1/track",
}
`;

exports[`creates a leaf when mime = application/json;something for req and res 1`] = `
{
  "methods": {
    "POST": {
      "queryParameters": {
        "properties": {
          "alt": {
            "type": "string",
          },
          "key": {
            "type": "string",
          },
        },
        "required": [
          "alt",
          "key",
        ],
        "type": "object",
      },
      "request": {
        "application/json;something": {
          "body": {
            "properties": {
              "test": {
                "type": "string",
              },
            },
            "required": [
              "test",
            ],
            "type": "object",
          },
        },
      },
      "requestHeaders": {
        "properties": {
          "anonymousid": {
            "type": "string",
          },
        },
        "required": [
          "anonymousid",
        ],
        "type": "object",
      },
      "response": {
        "200": {
          "application/json": {
            "body": {
              "properties": {
                "test": {
                  "type": "integer",
                },
              },
              "required": [
                "test",
              ],
              "type": "object",
            },
          },
        },
      },
      "responseHeaders": undefined,
    },
  },
  "pathname": "/v1/track",
}
`;

exports[`creates a leaf when request mime = application/x-www-form-urlencoded 1`] = `
{
  "methods": {
    "POST": {
      "queryParameters": undefined,
      "request": {
        "application/x-www-form-urlencoded": {
          "body": {
            "properties": {
              "comments": {
                "type": "string",
              },
              "custemail": {
                "type": "string",
              },
              "custname": {
                "type": "string",
              },
              "custtel": {
                "type": "string",
              },
              "delivery": {
                "type": "string",
              },
              "size": {
                "type": "string",
              },
              "topping": {
                "type": "string",
              },
            },
            "required": [
              "custname",
              "custtel",
              "custemail",
              "size",
              "topping",
              "delivery",
              "comments",
            ],
            "type": "object",
          },
        },
      },
      "requestHeaders": undefined,
      "response": {
        "200": {
          "application/json": {
            "body": {
              "properties": {
                "test": {
                  "type": "integer",
                },
              },
              "required": [
                "test",
              ],
              "type": "object",
            },
          },
        },
      },
      "responseHeaders": undefined,
    },
  },
  "pathname": "/post",
}
`;

exports[`creates a leaf when request mime starts with application/x-www-form-urlencoded;charset=UTF-8 1`] = `
{
  "methods": {
    "POST": {
      "queryParameters": undefined,
      "request": {
        "application/x-www-form-urlencoded;charset=UTF-8": {
          "body": {
            "properties": {
              "comments": {
                "type": "string",
              },
              "custemail": {
                "type": "string",
              },
              "custname": {
                "type": "string",
              },
              "custtel": {
                "type": "string",
              },
              "delivery": {
                "type": "string",
              },
              "size": {
                "type": "string",
              },
              "topping": {
                "type": "string",
              },
            },
            "required": [
              "custname",
              "custtel",
              "custemail",
              "size",
              "topping",
              "delivery",
              "comments",
            ],
            "type": "object",
          },
        },
      },
      "requestHeaders": undefined,
      "response": {
        "200": {
          "application/json": {
            "body": {
              "properties": {
                "test": {
                  "type": "integer",
                },
              },
              "required": [
                "test",
              ],
              "type": "object",
            },
          },
        },
      },
      "responseHeaders": undefined,
    },
  },
  "pathname": "/post",
}
`;
